### ISSUE TITLE
Does SchXSLT support schematrons embedded in RelaxNG and XMLSchema?

### ISSUE THREAD

**Issue by GerbenAbbink** [schxslt/schxslt#80]
Schematron schemas can be embedded in other languages such as RelaxNG and XMLSchema.
See http://www.topologi.com/resources/schtrn_xsd_paper.html.

Does SchXSLT support that?

**Comment by dmj** [comment]
Not directly, yet. The way it would be done is described in the paper you linked to (thanks! didn't know this one): There will be one or more stylesheets that extract a Schematron from XSD or RNG.

I make this a feature request and will target it with the 1.4 release of SchXslt.

**Comment by dmj** [comment]
Things to consider

- [ ] Obtain namespace bindings from the host language
- [ ] Perform includes for the host file, too

**Comment by dmj** [comment]
Maybe put this in a separate repository? Pro: Versioning independent of SchXslt

**Comment by GerbenAbbink** [comment]
Perfect!

As an example, here is the XSLT stylesheet (see attachment) that we use as
a preprocessor for Rick Jellife's Schematron implementation. I can't
remember if we made it ourselves or that we found it on the Internet. I
don't think it is part of Rick Jellife's implementation.

- Gerben Abbink
  XMLBlueprint.com

**Comment by dmj** [comment]
It seems the attached XSLT stylesheet had been stripped by Github. Could I ask you do put it e.g. in a Gist (https://gist.github.com) for me to download?

**Comment by GerbenAbbink** [comment]
Sorry about that.

I have made this public gist:
https://gist.github.com/GerbenAbbink/a265f628c073f3c8b817783a71dce621. Can
you read it?

- Gerben Abbink
  XMLBlueprint.com

**Comment by dmj** [comment]
Yes, thanks.

**Comment by dmj** [comment]
I'm no longer sure about if this belongs to the core of SchXslt. Removed it from the 1.5 release.

**Comment by adunning** [comment]
It would be an immense help to have support in schxslt for validating a file using rules embedded in an RNG schema. This is the normal way of distributing a schema in TEI files (where Schematron is very important) and I think in DocBook as well.

The lack is an obvious way to accomplish this means that there still isn't support for Schematron in the usual XML language servers. Common editors such as VS Code still only cover RelaxNG, and this is causing problems for contributors who do not realize their editor is not checking all the rules.

I am struggling to work out how to validate a file against an RNG file using schxslt. I know it involves using Saxon to extract the rules but I cannot find an XSL file that works. The one that this thread links to produces something close but the resulting .sch file is invalid. Does anyone have a suggestion?

**Comment by dmj** [comment]
@adunning I agree that having SchXslt use embedded schematron would be very helpful. As you have discovered yourself, a script that works for one case does not necessary work for another. 

That said: Could I ask you to try this one?

https://gist.github.com/dmj/8e099668fea614c7e2ad843aa97326a6

It's my attempt to write a transformation that makes as little assumptions about the host language as possible.

**Comment by adunning** [comment]
Thank you! I'm using the [standard TEI schemas](https://tei-c.org/release/xml/tei/custom/schema/relaxng/). It's not quite working and I wonder if I am using the right commands. I feel rather silly asking, but would you mind an example of how one would use Saxon with SchXslt in such a case, please?

(If an example is helpful, one file I'm looking to validate is <https://github.com/bodleian/medieval-mss/blob/master/works.xml> against <https://tei-c.org/release/xml/tei/custom/schema/relaxng/tei_ms.rng>.)

**Comment by dmj** [comment]
Step 1: Extract the Schematron from the RELAX NG.

```
java -jar /path/to/saxon.jar -xsl:extract-schematron.xsl -o:tei_ms.sch tei_ms.rng
```

Step 2: Validate work.xml against the extracted Schematron. The easiest way is with the SchXslt command line application:

https://github.com/schxslt/schxslt/releases/download/v1.9.5/schxslt-cli.jar

```
java -jar schxslt-cli.jar -d works.xml -s tei_ms.sch -v
```
